flake8_trio/visitors/visitor91x.py

```diff
@@ -7,37 +7,89 @@
 
 from __future__ import annotations
 
-import ast
+from dataclasses import dataclass, field
 from typing import Any
 
+import libcst as cst
+import libcst.matchers as m
+from libcst.metadata import PositionProvider
+
 from ..base import Statement
-from .flake8triovisitor import Flake8TrioVisitor
+from .flake8triovisitor import Flake8TrioVisitor_cst
 from .helpers import (
+    cst_literal_eval,
     disabled_by_default,
-    error_class,
-    fnmatch_qualified_name,
-    has_decorator,
-    iter_guaranteed_once,
+    error_class_cst,
+    fnmatch_qualified_name_cst,
+    func_has_decorator,
+    iter_guaranteed_once_cst,
 )
 
+ARTIFICIAL_STATEMENT = Statement("artificial", -1)
+
 
-# used in 910/911
-def empty_body(body: list[ast.stmt]) -> bool:
+def func_empty_body(node: cst.FunctionDef) -> bool:
     # Does the function body consist solely of `pass`, `...`, and (doc)string literals?
-    return all(
-        isinstance(stmt, ast.Pass)
-        or (
-            isinstance(stmt, ast.Expr)
-            and isinstance(stmt.value, ast.Constant)
-            and (stmt.value.value is Ellipsis or isinstance(stmt.value.value, str))
-        )
-        for stmt in body
+    return m.matches(
+        node.body,
+        m.IndentedBlock(
+            body=[
+                m.ZeroOrMore(
+                    m.SimpleStatementLine(
+                        body=[
+                            m.ZeroOrMore(
+                                m.Pass() | m.Expr(m.Ellipsis() | m.SimpleString())
+                            )
+                        ]
+                    )
+                )
+            ]
+        ),
+    )
+
+
+@dataclass
+class LoopState:
```

Comment:
So many state variables I decided to toss them into dataclasses. It mostly just makes save_state/restore_state stuff easier and less prone to missing a variable.